shared/Microsoft.Extensions.Buffers.Sources/MemoryPoolBlock.Debug.cs

```diff
@@ -0,0 +1,127 @@
+
+#if DEBUG
+
+using System.Threading;
+using System.Diagnostics;
+
+namespace System.Buffers
+{
+    /// <summary>
+    /// Block tracking object used by the byte buffer memory pool. A slab is a large allocation which is divided into smaller blocks. The
+    /// individual blocks are then treated as independent array segments.
+    /// </summary>
+    internal sealed class MemoryPoolBlock : MemoryManager<byte>
+    {
+        private readonly int _offset;
+        private readonly int _length;
+
+        private int _pinCount;
+
+        /// <summary>
+        /// This object cannot be instantiated outside of the static Create method
+        /// </summary>
+        internal MemoryPoolBlock(SlabMemoryPool pool, MemoryPoolSlab slab, int offset, int length)
+        {
+            _offset = offset;
+            _length = length;
+
+            Pool = pool;
+            Slab = slab;
+            Length = slab.Array.Length;
+        }
+
+        /// <summary>
+        /// Back-reference to the memory pool which this block was allocated from. It may only be returned to this pool.
+        /// </summary>
+        public SlabMemoryPool Pool { get; }
+
+        /// <summary>
+        /// Back-reference to the slab from which this block was taken, or null if it is one-time-use memory.
+        /// </summary>
+        public MemoryPoolSlab Slab { get; }
+
+        public override int Length { get; }
+
+        public override Memory<byte> Memory
+        {
+            get
+            {
+                if (!Slab.IsActive) ThrowHelper.ThrowObjectDisposedException(ExceptionArgument.MemoryPoolBlock);
+                return new Memory<byte>(this, _offset, _length);
+            }
+        }
+
+
+#if BLOCK_LEASE_TRACKING
+        public bool IsLeased { get; set; }
+        public string Leaser { get; set; }
+#endif
+
+        ~MemoryPoolBlock()
+        {
+            if (Slab != null && Slab.IsActive)
+            {
+               Debug.Assert(false, $"{Environment.NewLine}{Environment.NewLine}*** Block being garbage collected instead of returned to pool" +
+#if BLOCK_LEASE_TRACKING
+                   $": {Leaser}" +
+#endif
+                   $" ***{ Environment.NewLine}");
+
+                // Need to make a new object because this one is being finalized
+                Pool.Return(new MemoryPoolBlock(Pool, Slab, _offset, _length));
+            }
+        }
+
+        protected override void Dispose(bool disposing)
+        {
+            if (!Slab.IsActive) ThrowHelper.ThrowObjectDisposedException(ExceptionArgument.MemoryPoolBlock);
+
+            if (Volatile.Read(ref _pinCount) > 0)
+            {
+                ThrowHelper.ThrowInvalidOperationException_ReturningPinnedBlock();
+            }
+
+            Pool.Return(this);
+        }
+
+        public override Span<byte> GetSpan() => new Span<byte>(Slab.Array);
+
+        public override MemoryHandle Pin(int byteOffset = 0)
```

Comment:
nit: This is elementIndex now (though in this case, since size of T == 1, it is the same.